src/NHibernate/Hql/Ast/ANTLR/Tree/BinaryLogicOperatorNode.cs

```diff
@@ -277,25 +278,29 @@ private void ProcessMetaTypeDiscriminatorIfNecessary(IASTNode lhs, IASTNode rhs)
 				return;
 			}
 
-			var lhsNodeMetaType = lhsNode.DataType as MetaType;
-			if (lhsNodeMetaType != null)
+			if (lhsNode.DataType is MetaType lhsNodeMetaType)
```

Comment:
Having a quick look at it, it seems the same could be done with `ClassMetaType`, which should allow supporting the `Class` mapping case. Or am I missing something?
(Maybe `ClassMetaType` should even extend `MetaType`, with a hard-coded in base constructor call `StringType` base type.)

Reply:
Well, the HQL test is invalid for that case, because `class` meta-type asks for the full class name.

By the way there is a workaround for all HQL cases: using parameters. It seems this was considered as a known limitation of the `any` mapping at some point, see [this Jira comment](https://nhibernate.jira.com/browse/NH-1403?focusedCommentId=16651&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16651).